src/v/datalake/record_multiplexer.cc

```diff
@@ -28,32 +32,41 @@ record_multiplexer::operator()(model::record_batch batch) {
     }
     auto first_timestamp = batch.header().first_timestamp.value();
     auto base_offset = static_cast<int64_t>(batch.base_offset());
-    co_await batch.for_each_record_async(
-      [first_timestamp, base_offset, this](
-        model::record record) -> ss::future<> {
-          iobuf key = record.release_key();
-          iobuf val = record.release_value();
-          // *1000: Redpanda timestamps are milliseconds. Iceberg uses
-          // microseconds.
-          int64_t timestamp = (first_timestamp + record.timestamp_delta())
-                              * 1000;
-          int64_t offset = static_cast<int64_t>(base_offset)
-                           + record.offset_delta();
-          int64_t estimated_size = key.size_bytes() + val.size_bytes() + 16;
+    auto it = model::record_batch_iterator::create(batch);
+    while (it.has_next()) {
+        auto record = it.next();
+        iobuf key = record.release_key();
+        iobuf val = record.release_value();
+        // *1000: Redpanda timestamps are milliseconds. Iceberg uses
+        // microseconds.
+        int64_t timestamp = (first_timestamp + record.timestamp_delta()) * 1000;
+        int64_t offset = static_cast<int64_t>(base_offset)
+                         + record.offset_delta();
+        int64_t estimated_size = key.size_bytes() + val.size_bytes() + 16;
 
-          // Translate the record
-          auto& translator = get_translator();
-          iceberg::struct_value data = translator.translate_event(
-            std::move(key), std::move(val), timestamp, offset);
-          // Send it to the writer
-          auto& writer = get_writer();
-          co_await writer.add_data_struct(std::move(data), estimated_size);
-      });
+        // Translate the record
+        auto& translator = get_translator();
+        iceberg::struct_value data = translator.translate_event(
+          std::move(key), std::move(val), timestamp, offset);
+        // Send it to the writer
+        auto& writer = get_writer();
+        data_writer_error writer_status
+          = writer.add_data_struct(std::move(data), estimated_size).get();
```

Comment:
it seems `co_await` is missing here

Reply:
it's not in a thread context, from what i can tell. so it's not crashing because the future is ready.